rewriter/src/main/java/com/facebook/coresql/rewriter/ApproxPercentileRewriter.java

```diff
@@ -0,0 +1,182 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.coresql.rewriter;
+
+import com.facebook.coresql.parser.AstNode;
+import com.facebook.coresql.parser.FunctionCall;
+import com.facebook.coresql.parser.SqlParserDefaultVisitor;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ListMultimap;
+
+import java.util.Formatter;
+import java.util.Optional;
+
+import static com.facebook.coresql.parser.SqlParserTreeConstants.JJTARGUMENTLIST;
+import static com.facebook.coresql.parser.SqlParserTreeConstants.JJTIDENTIFIER;
+import static com.facebook.coresql.parser.SqlParserTreeConstants.JJTUNSIGNEDNUMERICLITERAL;
+import static java.util.Objects.requireNonNull;
+
+public class ApproxPercentileRewriter
+        extends Rewriter
+{
+    private static final int ZERO_INDEXING_OFFSET = 1;
+    private static final String REPLACEMENT_FORMAT = " APPROX_PERCENTILE(%s, ARRAY%s)[%d]";
+    private static final String REWRITE_NAME = "Multiple APPROX PERCENTILE with same first arg and literal second arg";
+    private final AstNode root;
+    private final ListMultimap<String, AstNode> firstArgToApproxPercentileNode; // A map of String to the APPROX_PERCENTILE nodes with that String as their first argument
+    private final ListMultimap<String, String> firstArgToPercentiles; // A map of String to a list of the percentiles of the APPROX_PERCENTILE nodes that have that String as their first argument
+
+    public ApproxPercentileRewriter(AstNode root)
+    {
+        this.root = requireNonNull(root, "AST passed to rewriter was null");
+        ApproxPercentilePatternMatcherResult patternMatcherResult = new ApproxPercentilePatternMatcher(root).matchPattern();
+        this.firstArgToApproxPercentileNode = patternMatcherResult.getFirstArgToApproxPercentileNode();
+        this.firstArgToPercentiles = patternMatcherResult.getFirstArgToPercentiles();
+    }
+
+    @Override
+    public Optional<RewriteResult> rewrite()
+    {
+        if (!approxPercentileRewritePatternIsPresent()) {
+            return Optional.empty();
+        }
+        String rewrittenSql = unparse(root, this);
+        return Optional.of(new RewriteResult(REWRITE_NAME, rewrittenSql));
+    }
+
+    @Override
+    public void visit(FunctionCall node, Void data)
+    {
+        if (canApplyApproxPercentileRewrite(node)) {
+            applyApproxPercentileRewrite(node);
+        }
+        else {
+            defaultVisit(node, data);
+        }
+    }
+
+    private boolean approxPercentileRewritePatternIsPresent()
+    {
+        return firstArgToApproxPercentileNode.keySet().stream()
+                .anyMatch(key -> firstArgToApproxPercentileNode.get(key).size() >= 2);
+    }
+
+    private boolean canApplyApproxPercentileRewrite(AstNode node)
```

Comment:
It should be more like: needsApproxPercentileRewrite

Reply:
Done

Reply:
Done.